README.pod
=pod

=encoding utf-8

=head1 NAME

Dist::Zilla::PluginBundle::Author::BBYRD - DZIL Author Bundle for BBYRD

=head1 SYNOPSIS

    ; Very similar to...
 
    [ReportPhase]
 
    ; Makefile.PL maker
    [MakeMaker]
 
    [ModuleShareDirs]
    Dist::Zilla::MintingProfile::Author::BBYRD = profiles
 
    [Git::NextVersion]
    first_version = 0.90
 
    [Git::GatherDir]
 
    ; File modifiers
    [OurPkgVersion]
    [PodWeaver]
    config_plugin = @Author::BBYRD
 
    ; File pruners
    [PruneCruft]
 
    ; Extra file creation
    [GitFmtChanges]
    [ManifestSkip]
    [Manifest]
    [License]
    [ReadmeAnyFromPod / ReadmePodInRoot]    ; Pod README for Root (for GitHub, etc.)
    [ReadmeAnyFromPod / ReadmeTextInBuild]  ; Text README for Build
    [ReadmeAnyFromPod / ReadmeHTMLInBuild]  ; HTML README for Build (never POD, so it doesn't get installed)
    [InstallGuide]
    [ExecDir]
 
    ; t/* tests
    [Test::Compile]
 
    ; POD tests
    [PodCoverageTests]
    [PodSyntaxTests]
    ;[Test::PodSpelling]  ; Win32 install problems
 
    ; Other xt/* tests
    [RunExtraTests]
    ;[MetaTests]  ; until Test::CPAN::Meta supports 2.0
    [NoTabsTests]
    [Test::EOL]
    trailing_whitespace = 0
 
    [Test::CPAN::Meta::JSON]
    [Test::CheckDeps]
    [Test::Portability]
    ;[Test::Pod::LinkCheck]  ; Both of these are borked...  
    ;[Test::Pod::No404s]     ; ...I really need to create my own
    [Test::Synopsis]
    [Test::MinimumVersion]
    [ReportVersions::Tiny]
    [Test::CheckManifest]
    [Test::DistManifest]
    [Test::Version]
 
    ; Prereqs
    [@Prereqs]
    minimum_perl = 5.10.1
 
    [CheckPrereqsIndexed]
 
    ; META maintenance
    [MetaConfig]
    [MetaJSON]
    [MetaYAML]
 
    [MetaNoIndex]
    directory = t
    directory = xt
    directory = examples
    directory = corpus
 
    [MetaProvides::Package]
    meta_noindex = 1        ; respect prior no_index directives
 
    [MetaResourcesFromGit]
    x_irc          = irc://irc.perl.org/#distzilla
    bugtracker.web = https://github.com/%a/%r/issues
 
    [ContributorsFromGit]
 
    ; Post-build plugins
    [CopyFilesFromBuild]
    move = .gitignore
    move = README.pod
 
    ; Post-build Git plugins
    [TravisYML]
    test_min_deps = 1
 
    [Git::CheckFor::CorrectBranch]
    [Git::CommitBuild]
    branch = 
    release_branch = build/%b
    release_message = Release build of v%v (on %b)
 
    [@Git]
    allow_dirty = dist.ini
    allow_dirty = .travis.yml
    allow_dirty = README.pod
    changelog =
    commit_msg = Release v%v
    push_to = origin master:master
    push_to = origin build/master:build/master
 
    [GitHub::Update]
    metacpan = 1
 
    [TestRelease]
    [ConfirmRelease]
    [UploadToCPAN]
    [InstallRelease]
    [Clean]
 
    ; sanity deps
    ; authordep autovivification
    ; authordep indirect
    ; authordep multidimensional

=head1 DESCRIPTION

L<I frelling hate these things|sanity>, but several releases in, I found myself
needing to keep my C<<< dist.ini >>> stuff in sync, which requires a single module to
bind them to.

=head1 NAMING SCHEME

I'm a strong believer in structured order in the chaos that is the CPAN
namespace.  There's enough cruft in CPAN, with all of the forked modules, 
legacy stuff that should have been removed 10 years ago, and confusion over
which modules are available vs. which ones actually work.  (Which all stem
from the same base problem, so I'm almost repeating myself...)

Like I said, I hate writing these personalized modules on CPAN.  I even bantered
around the idea of using L<MetaCPAN's author JSON input|https://github.com/SineSwiper/Dist-Zilla-PluginBundle-BeLike-You/blob/master/BeLike-You.pod>
to store the plugin data.  However, keeping the Author plugins separated from the
real PluginBundles is a step in the right direction.  See
L<KENTNL's comments on the Author namespace|Dist::Zilla::PluginBundle::Author::KENTNL/NAMING-SCHEME>
for more information.

=head1 CAVEATS

This uses L<Dist::Zilla::Role::PluginBundle::Merged>, so all of the plugins'
arguments are available, using Merged's rules.  Special care should be
made with arguments that might not be unique with other plugins.  (Eventually,
I'll throw these into C<<< config_rename >>>.)

If this is a problem, you might want to consider using L<@Filter|Dist::Zilla::PluginBundle::Filter>.

One exception is C<<< x_irc >>>, which is detected and passed to L<MetaResourcesFromGit|Dist::Zilla::Plugin::MetaResourcesFromGit>
properly.

=head1 SEE ALSO

In building my ultimate C<<< dist.ini >>> file, I did a bunch of research on which
modules to cram in here.  As a result, this is a pretty large set of plugins,
but that's exactly how I like my DZIL.  Feel free to research the modules
listed here, as there's a bunch of good modules that you might want to include
in your own C<<< dist.ini >>> andE<sol>or Author bundle.

Also, here's my C<<< profile.ini >>>, if you're interested:

    [TemplateModule/:DefaultModuleMaker]
    template = Module.pm
 
    [DistINI]
    append_file = plugins.ini
 
    [GatherDir::Template]
    root = skel
 
    [GenerateFile / Generate-.gitignore]
    filename = .gitignore
    is_template = 1
    content = MANIFEST
    content = MANIFEST.bak
    content = Makefile
    content = Makefile.old
    content = Build
    content = Build.bat
    content = META.*
    content = MYMETA.*
    content = .build/
    content = _build/
    content = blib/
    content = inc/
    content = .lwpcookies
    content = .last_cover_stats
    content = nytprof.out
    content = pod2htm*.tmp
    content = pm_to_blib
    content = {{$dist->name}}-*
    content = {{$dist->name}}-*.tar.gz
 
    [Git::Init]
    commit_message = Initial commit
 
    [GitHub::Create]

=head1 AVAILABILITY

The project homepage is L<https://github.com/SineSwiper/Dist-Zilla-PluginBundle-Author-BBYRD/wiki>.

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see L<https://metacpan.org/module/Dist::Zilla::PluginBundle::Author::BBYRD/>.

=for :stopwords cpan testmatrix url annocpan anno bugtracker rt cpants kwalitee diff irc mailto metadata placeholders metacpan

=head1 SUPPORT

=head2 Internet Relay Chat

You can get live help by using IRC ( Internet Relay Chat ). If you don't know what IRC is,
please read this excellent guide: L<http://en.wikipedia.org/wiki/Internet_Relay_Chat>. Please
be courteous and patient when talking to us, as we might be busy or sleeping! You can join
those networks/channels and get help:

=over 4

=item *

irc.perl.org

You can connect to the server at 'irc.perl.org' and join this channel: #distzilla then talk to this person for help: SineSwiper.

=back

=head2 Bugs / Feature Requests

Please report any bugs or feature requests via L<https://github.com/SineSwiper/Dist-Zilla-PluginBundle-Author-BBYRD/issues>.

=head1 AUTHOR

Brendan Byrd <BBYRD@CPAN.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by Brendan Byrd.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)



CHANGES
Revision History for Dist-Zilla-PluginBundle-Author-BBYRD
=========================================================

v0.95 2013-05-15
----------------

commit 6783709d3fd16deff8b705f73f0f1a9e6e41875f
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Sat Apr 27 12:48:46 2013 -0400

    Fix new dependency changes

commit 3011444e28a23a205c041647b7abfc84693b763d
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Sat Apr 27 10:16:50 2013 -0400

    Add Author minting profile

commit 1ec6fd43a7d1929ff55ad4425895baac160728f0
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Sat Apr 27 10:10:41 2013 -0400

    Remove IRC room name from Support section
    Remove Test::UseAllModules (redundant with Test::Compile)
    Fix README.pod to not end up in build (was installing in lib/)
    Add support for Contributors (via Git)

commit 45fe424bc6403729e98a0796c48e59896fa222d6
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Thu Apr 25 09:45:28 2013 -0400

    Be more explicit with push master:master
    Fix CommitBuild to not create random builds
    Fix .travis.yml to run through deps faster

v0.94 2013-04-22
----------------

commit 8ae22fcee32f6b8ddebcb842d9157563e2845f2e
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Mon Apr 22 16:27:31 2013 -0400

    Update .travis.yml with new DZIL:P:TCI defaults

commit d99b8f45d866aa709108ddbc78c07a6b93257a70
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Mon Apr 22 10:45:22 2013 -0400

    Add new Pod::Weaver::PluginBundle::Author::BBYRD

v0.93 2012-10-17
----------------

commit b395ec4299c1ae2c0aa748f43c520a51f46fe241
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Wed Oct 17 13:58:00 2012 -0400

    Release v0.93

commit bcc0084b89b69e7a846611c198e96c6ade71fcb1
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Tue Oct 16 11:56:47 2012 -0400

    Add [MetaYAML] back in

v0.92 2012-10-02
----------------

commit 1325030d6e61ccbbc650b7d3f7f5c3b478cd5e8f
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Tue Oct 2 11:30:59 2012 -0400

    Release v0.92

commit e89f92b1e221d25aa502af361081c82e3280ee2f
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Tue Oct 2 11:29:24 2012 -0400

    Fix MetaResourcesFromGit call

v0.91 2012-10-02
----------------

commit 649910d9a8bbfb5131f008a7d6f389be34329c60
Author: Brendan Byrd <Perl@ResonatorSoft.org>
Date:   Tue Oct 2 08:39:23 2012 -0400

    Fix POD

===================================
End of changes in the last 365 days
===================================


lib/Pod/Weaver/PluginBundle/Author/BBYRD.pm
package Pod::Weaver::PluginBundle::Author::BBYRD;

our $VERSION = '0.95'; # VERSION
# ABSTRACT: Pod::Weaver Author Bundle for BBYRD

use sanity;
 
use Pod::Weaver 3.101635; # fixed ABSTRACT scanning
use Pod::Weaver::Config::Assembler;
 
# Dependencies
use Pod::Weaver::Plugin::WikiDoc ();
use Pod::Weaver::Plugin::Encoding ();
use Pod::Weaver::Section::Availability ();
use Pod::Elemental::Transformer::List 0.101620 ();
use Pod::Weaver::Section::Support 1.001        ();
use Pod::Weaver::Section::Contributors ();
 
sub _exp { Pod::Weaver::Config::Assembler->expand_package( $_[0] ) }

### TODO: Consider a simplified version of this a la DZIL:R:PB:Merged?
sub mvp_bundle_config {
   my @plugins;
   push @plugins, (
      # [-Encoding]
      # [-WikiDoc]
      # [@CorePrep]
      #  
      # [Name]

      [ '@Author::BBYRD/Encoding', _exp('-Encoding'), {} ],
      [ '@Author::BBYRD/WikiDoc',  _exp('-WikiDoc'),  {} ],
      [ '@Author::BBYRD/CorePrep', _exp('@CorePrep'), {} ],
      [ '@Author::BBYRD/Name',     _exp('Name'),      {} ],
 
      # [Region / prelude]
      #  
      # [Generic / SYNOPSIS]
      # [Generic / DESCRIPTION]
      # [Generic / OVERVIEW]

      [ '@Author::BBYRD/Prelude',     _exp('Region'),  { region_name => 'prelude' } ],
      [ '@Author::BBYRD/Synopsis',    _exp('Generic'), { header      => 'SYNOPSIS' } ],
      [ '@Author::BBYRD/Description', _exp('Generic'), { header      => 'DESCRIPTION' } ],
      [ '@Author::BBYRD/Overview',    _exp('Generic'), { header      => 'OVERVIEW' } ],
   );
 
   foreach my $plugin (
      # [Collect / ATTRIBUTES]
      # command = attr
      #  
      # [Collect / METHODS]
      # command = method
      #  
      # [Collect / FUNCTIONS]
      # command = func

      [ 'Attributes',   _exp('Collect'), { command => 'attr' } ],
      [ 'Methods',      _exp('Collect'), { command => 'method' } ],
      [ 'Functions',    _exp('Collect'), { command => 'func' } ],
     )
   {
       $plugin->[2]{header} = uc $plugin->[0];
       push @plugins, $plugin;
   }

   push @plugins, (
      # [Leftovers]
      #  
      # [Region / postlude]
      #  
      # [Availability]
      
      [ '@Author::BBYRD/Leftovers',    _exp('Leftovers'),    {} ],
      [ '@Author::BBYRD/postlude',     _exp('Region'),       { region_name => 'postlude' } ],
      [ '@Author::BBYRD/Availability', _exp('Availability'), {} ],
      
      # [Support]
      # perldoc = 0
      # websites = none
      # repository_link = none
      # bugs = metadata
      # bugs_content = Please report any bugs or feature requests via {WEB}.
      # irc = irc.perl.org, #distzilla, SineSwiper
      
      [ '@Author::BBYRD/Support', _exp('Support'), {
         perldoc            => 0,
         websites           => 'none',
         repository_link    => 'none',
         bugs               => 'metadata',
         bugs_content       => 'Please report any bugs or feature requests via {WEB}.',
         ### XXX: Use a DZIL stash to store the IRC channel? ###
         irc                => 'irc.perl.org, SineSwiper',
      } ],
      
      # [Authors]
      # [Contributors]
      # [Legal]
      #  
      # [-Transformer]
      # transformer = List

      [ '@Author::BBYRD/Authors',      _exp('Authors'),      {} ],
      [ '@Author::BBYRD/Contributors', _exp('Contributors'), {} ],
      [ '@Author::BBYRD/Legal',        _exp('Legal'),        {} ],
      [ '@Author::BBYRD/List',         _exp('-Transformer'), { 'transformer' => 'List' } ],
   );
 
   return @plugins;
}

42;

__END__

=pod

=encoding utf-8

=head1 NAME

Pod::Weaver::PluginBundle::Author::BBYRD - Pod::Weaver Author Bundle for BBYRD

=head1 SYNOPSIS

    ; Very similar to...
 
    [-Encoding]
    [-WikiDoc]
    [@CorePrep]
 
    [Name]
 
    [Region / prelude]
 
    [Generic / SYNOPSIS]
    [Generic / DESCRIPTION]
    [Generic / OVERVIEW]
 
    [Collect / ATTRIBUTES]
    command = attr
 
    [Collect / METHODS]
    command = method
 
    [Collect / FUNCTIONS]
    command = func
 
    [Leftovers]
 
    [Region / postlude]
 
    [Availability]
    [Support]
    perldoc = 0
    websites = none
    repository_link = none
    bugs = metadata
    bugs_content = Please report any bugs or feature requests via {WEB}.
    irc = irc.perl.org, SineSwiper
 
    [Authors]
    [Contributors]
    [Legal]
 
    [-Transformer]
    transformer = List
 
    ; PodWeaver deps
    ; authordep Pod::Weaver::Plugin::WikiDoc
    ; authordep Pod::Weaver::Plugin::Encoding
    ; authordep Pod::Weaver::Section::Availability
    ; authordep Pod::Weaver::Section::Support
    ; authordep Pod::Elemental::Transformer::List

=head1 DESCRIPTION

Like the DZIL one, this is a personalized Author bundle for my Pod::Weaver
configuration.

=head1 NAMING SCHEME

I'm a strong believer in structured order in the chaos that is the CPAN
namespace.  There's enough cruft in CPAN, with all of the forked modules, 
legacy stuff that should have been removed 10 years ago, and confusion over
which modules are available vs. which ones actually work.  (Which all stem
from the same base problem, so I'm almost repeating myself...)

Like I said, I hate writing these personalized modules on CPAN.  I even bantered
around the idea of using L<MetaCPAN's author JSON input|https://github.com/SineSwiper/Dist-Zilla-PluginBundle-BeLike-You/blob/master/BeLike-You.pod>
to store the plugin data.  However, keeping the Author plugins separated from the
real PluginBundles is a step in the right direction.  See
L<KENTNL's comments on the Author namespace|Dist::Zilla::PluginBundle::Author::KENTNL/NAMING-SCHEME>
for more information.

=head1 AVAILABILITY

The project homepage is L<https://github.com/SineSwiper/Dist-Zilla-PluginBundle-Author-BBYRD/wiki>.

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see L<https://metacpan.org/module/Dist::Zilla::PluginBundle::Author::BBYRD/>.

=head1 AUTHOR

Brendan Byrd <BBYRD@CPAN.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2013 by Brendan Byrd.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


Makefile.PL

use strict;
use warnings;

use 5.010001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "DZIL Author Bundle for BBYRD",
  "AUTHOR" => "Brendan Byrd <BBYRD\@CPAN.org>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {},
  "DISTNAME" => "Dist-Zilla-PluginBundle-Author-BBYRD",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "Dist::Zilla::PluginBundle::Author::BBYRD",
  "PREREQ_PM" => {
    "Dist::Zilla" => "4.300034",
    "Dist::Zilla::Plugin::CheckExtraTests" => "0.011",
    "Dist::Zilla::Plugin::CheckPrereqsIndexed" => "0.008",
    "Dist::Zilla::Plugin::Clean" => "0.07",
    "Dist::Zilla::Plugin::ContributorsFromGit" => "0.006",
    "Dist::Zilla::Plugin::CopyFilesFromBuild" => "0.103510",
    "Dist::Zilla::Plugin::GitFmtChanges" => "0.005",
    "Dist::Zilla::Plugin::InstallGuide" => "1.200000",
    "Dist::Zilla::Plugin::InstallRelease" => "0.008",
    "Dist::Zilla::Plugin::MetaProvides::Package" => "1.14000002",
    "Dist::Zilla::Plugin::MetaResourcesFromGit" => "1.103620",
    "Dist::Zilla::Plugin::NoTabsTests" => "0.01",
    "Dist::Zilla::Plugin::OurPkgVersion" => "0.004000",
    "Dist::Zilla::Plugin::PodWeaver" => "3.101641",
    "Dist::Zilla::Plugin::ReadmeAnyFromPod" => "0.120120",
    "Dist::Zilla::Plugin::ReportPhase" => 0,
    "Dist::Zilla::Plugin::ReportVersions::Tiny" => "1.08",
    "Dist::Zilla::Plugin::Test::CPAN::Meta::JSON" => "0.003",
    "Dist::Zilla::Plugin::Test::CheckDeps" => "0.006",
    "Dist::Zilla::Plugin::Test::CheckManifest" => "0.04",
    "Dist::Zilla::Plugin::Test::Compile" => "2.002",
    "Dist::Zilla::Plugin::Test::DistManifest" => "2.000003",
    "Dist::Zilla::Plugin::Test::EOL" => "0.07",
    "Dist::Zilla::Plugin::Test::MinimumVersion" => "2.000004",
    "Dist::Zilla::Plugin::Test::Portability" => "2.000004",
    "Dist::Zilla::Plugin::Test::Synopsis" => "2.000003",
    "Dist::Zilla::Plugin::Test::Version" => "0.002004",
    "Dist::Zilla::Plugin::TravisYML" => "0.97",
    "Dist::Zilla::PluginBundle::Git" => "2.013",
    "Dist::Zilla::PluginBundle::Git::CheckFor" => "0.006",
    "Dist::Zilla::PluginBundle::GitHub" => "0.34",
    "Dist::Zilla::PluginBundle::Prereqs" => "0.92",
    "Dist::Zilla::Role::PluginBundle::Merged" => "0.92",
    "Dist::Zilla::Stash::PAUSE::Encrypted" => "0.003",
    "Moose" => "2.0802",
    "Pod::Coverage::TrustPod" => "0.100002",
    "Pod::Elemental::Transformer::List" => "0.101620",
    "Pod::Weaver" => "3.101638",
    "Pod::Weaver::Plugin::Encoding" => "0.01",
    "Pod::Weaver::Plugin::WikiDoc" => "0.093002",
    "Pod::Weaver::Section::Availability" => "1.20",
    "Pod::Weaver::Section::Contributors" => "0.005",
    "Pod::Weaver::Section::Support" => "1.001",
    "Test::CPAN::Meta::JSON" => "0.14",
    "autovivification" => "0.11",
    "indirect" => "0.29",
    "multidimensional" => "0.011",
    "sanity" => "0.95"
  },
  "TEST_REQUIRES" => {
    "Test::CheckDeps" => "0.002",
    "Test::UseAllModules" => "0.14"
  },
  "VERSION" => "0.95",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



